library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/SessionReplayRecorder.kt

```diff
@@ -0,0 +1,183 @@
+ */
+
+package com.datadog.android.sessionreplay
+
+import android.app.Application
+import android.os.Handler
+import android.os.Looper
+import android.view.Window
+import androidx.annotation.VisibleForTesting
+import com.datadog.android.sessionreplay.internal.LifecycleCallback
+import com.datadog.android.sessionreplay.internal.NoOpRecordCallback
+import com.datadog.android.sessionreplay.internal.RecordCallback
+import com.datadog.android.sessionreplay.internal.RecordWriter
+import com.datadog.android.sessionreplay.internal.SessionReplayLifecycleCallback
+import com.datadog.android.sessionreplay.internal.processor.RecordedDataProcessor
+import com.datadog.android.sessionreplay.internal.recorder.ComposedOptionSelectorDetector
+import com.datadog.android.sessionreplay.internal.recorder.DefaultOptionSelectorDetector
+import com.datadog.android.sessionreplay.internal.recorder.OptionSelectorDetector
+import com.datadog.android.sessionreplay.internal.recorder.SnapshotProducer
+import com.datadog.android.sessionreplay.internal.recorder.TreeViewTraversal
+import com.datadog.android.sessionreplay.internal.recorder.ViewOnDrawInterceptor
+import com.datadog.android.sessionreplay.internal.recorder.WindowCallbackInterceptor
+import com.datadog.android.sessionreplay.internal.recorder.WindowInspector
+import com.datadog.android.sessionreplay.internal.recorder.callback.OnWindowRefreshedCallback
+import com.datadog.android.sessionreplay.internal.recorder.mapper.MapperTypeWrapper
+import com.datadog.android.sessionreplay.internal.utils.RumContextProvider
+import com.datadog.android.sessionreplay.internal.utils.TimeProvider
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+internal class SessionReplayRecorder : OnWindowRefreshedCallback, Recorder {
+
+    private val appContext: Application
+    private val rumContextProvider: RumContextProvider
+    private val privacy: SessionReplayPrivacy
+    private val recordWriter: RecordWriter
+    private val timeProvider: TimeProvider
+    private val recordCallback: RecordCallback
+    private val customMappers: List<MapperTypeWrapper>
+    private val customOptionSelectorDetectors: List<OptionSelectorDetector>
+    private val windowInspector: WindowInspector
+    private val windowCallbackInterceptor: WindowCallbackInterceptor
+    private val sessionReplayLifecycleCallback: LifecycleCallback
+    private val processor: RecordedDataProcessor
+    private val viewOnDrawInterceptor: ViewOnDrawInterceptor
+    private val uiHandler: Handler
+
+    @Suppress("UnsafeThirdPartyFunctionCall") // workQueue can't be null
+    private val processorExecutorService = ThreadPoolExecutor(
+        CORE_DEFAULT_POOL_SIZE,
+        Runtime.getRuntime().availableProcessors(),
+        THREAD_POOL_MAX_KEEP_ALIVE_MS,
+        TimeUnit.MILLISECONDS,
+        LinkedBlockingDeque()
+    )
+    private var shouldRecord = false
```

Comment:
nope it will be only called from `MainThread` that's the reason why I am using the `uiHandler` inside to make sure of that. Also I think the `boolean` primitive is volatile by default but agree that in terms of atomicity and multi - thread modification an `AtomicBoolean` should be used but not in this case.

Reply:
In that case can you annotate the methods with `@MainThread` to make this explicit

Reply:
if `boolean` is thread-safe, then it wouldn't be an `AtomicBoolean` in the Java API :)

Reply:
one thing is to be volatile (which means its value will be the same in different threads and it will not be cached in the thread local memory) and the other thing is to be `Atomic` meaning that you will not update it in the same time from 2 threads and you can apply (update only if previous was x)